File: smoke_tests/perl/log_injection.pl
#!/usr/bin/perl
# Log Injection vulnerabilities in Perl

use strict;
use warnings;
use CGI;

my $cgi = CGI->new();

# Test 1: warn() with user input
sub vulnerable_warn {
    my $username = $cgi->param('username');
    # VULNERABLE: Log injection via warn
    warn "Login attempt for user: $username";
}

# Test 2: die() with user input
sub vulnerable_die {
    my $file = $cgi->param('file');
    # VULNERABLE: Log injection via die
    die "Cannot open file: $file" unless -e $file;
}

# Test 3: Carp::carp with user input
sub vulnerable_carp {
    use Carp;
    my $action = $cgi->param('action');
    # VULNERABLE: Log injection via carp
    carp "Invalid action requested: $action";
}

# Test 4: Carp::croak with user input
sub vulnerable_croak {
    use Carp;
    my $input = $cgi->param('input');
    # VULNERABLE: Log injection via croak
    croak "Validation failed for input: $input" unless $input =~ /^\d+$/;
}

# Test 5: Carp::cluck with user input
sub vulnerable_cluck {
    use Carp;
    my $param = $cgi->param('param');
    # VULNERABLE: Log injection via cluck
    cluck "Suspicious parameter value: $param";
}

# Test 6: Carp::confess with user input
sub vulnerable_confess {
    use Carp;
    my $error = $cgi->param('error');
    # VULNERABLE: Log injection via confess
    confess "Fatal error: $error";
}

# Test 7: print to STDERR
sub vulnerable_stderr {
    my $message = $cgi->param('message');
    # VULNERABLE: Log injection via STDERR
    print STDERR "Error: $message\n";
}

# Test 8: syslog with user input
sub vulnerable_syslog {
    use Sys::Syslog qw(:standard :macros);
    my $username = $cgi->param('username');
    openlog('myapp', 'ndelay,pid', LOG_LOCAL0);
    # VULNERABLE: Log injection via syslog
    syslog(LOG_INFO, "User $username logged in");
    closelog();
}

# Test 9: Log::Log4perl
sub vulnerable_log4perl {
    use Log::Log4perl;
    my $logger = Log::Log4perl->get_logger();
    my $user_agent = $ENV{HTTP_USER_AGENT};
    # VULNERABLE: Log injection via Log4perl
    $logger->info("Request from: $user_agent");
}

# Test 10: Log::Dispatch
sub vulnerable_log_dispatch {
    use Log::Dispatch;
    my $log = Log::Dispatch->new(outputs => [['Screen', min_level => 'info']]);
    my $path = $cgi->param('path');
    # VULNERABLE: Log injection via Log::Dispatch
    $log->info("Accessing path: $path");
}

# Test 11: File logging with user input
sub vulnerable_file_log {
    my $action = $cgi->param('action');
    my $user = $cgi->param('user');
    open(my $fh, '>>', '/var/log/myapp.log') or die;
    # VULNERABLE: Log injection via file
    print $fh "[" . localtime() . "] User $user performed action: $action\n";
    close($fh);
}

# Test 12: Apache log via print
sub vulnerable_apache_log {
    my $referer = $ENV{HTTP_REFERER};
    # VULNERABLE: Log injection in Apache error log
    print STDERR "[error] Invalid referer: $referer\n";
}

# Test 13: Log newline injection
sub vulnerable_newline_injection {
    my $data = $cgi->param('data');
    # VULNERABLE: Newline injection to forge log entries
    warn "Received data: $data";
    # Attacker can inject: "data\n[INFO] Admin logged in successfully"
}

# Test 14: JSON in logs
sub vulnerable_json_log {
    use JSON;
    my $input = $cgi->param('input');
    my $log_entry = encode_json({
        timestamp => time(),
        message => "Processed input",
        # VULNERABLE: User input in JSON log
        user_data => $input
    });
    print STDERR "$log_entry\n";
}

# Test 15: Audit logging
sub vulnerable_audit_log {
    my $dbh = get_dbh();
    my $action = $cgi->param('action');
    my $user_id = $cgi->cookie('user_id');
    # VULNERABLE: Log injection in database audit log
    $dbh->do("INSERT INTO audit_log (user_id, action, timestamp) VALUES (?, ?, NOW())",
        undef, $user_id, $action);
}

# Helper functions
sub get_dbh { return DBI->connect("dbi:SQLite:dbname=test.db"); }

1;

File: smoke_tests/csharp/log_injection.cs
// Log Injection vulnerabilities in C#
using System;
using System.IO;
using System.Diagnostics;
using System.Web.Mvc;
using Microsoft.Extensions.Logging;
using log4net;
using NLog;

namespace VulnerableApp
{
    public class LogInjectionController : Controller
    {
        private readonly ILog _log4net = LogManager.GetLogger(typeof(LogInjectionController));
        private readonly Logger _nlog = LogManager.GetCurrentClassLogger();
        private readonly ILogger<LogInjectionController> _msLogger;

        // Test 1: Console.WriteLine with user input
        public ActionResult ConsoleLog()
        {
            string username = Request.QueryString["username"];
            // VULNERABLE: Can inject newlines to forge log entries
            Console.WriteLine($"[INFO] User logged in: {username}");
            return Ok();
        }

        // Test 2: Debug.WriteLine
        public ActionResult DebugLog()
        {
            string action = Request.Form["action"];
            // VULNERABLE: User input in debug log
            Debug.WriteLine($"Action performed: {action}");
            return Ok();
        }

        // Test 3: File logging with user input
        public ActionResult FileLog()
        {
            string message = Request.Form["message"];
            // VULNERABLE: Message can contain CRLF
            File.AppendAllText("/var/log/app.log",
                $"[{DateTime.Now}] {message}\n");
            return Ok();
        }

        // Test 4: EventLog with user input
        public ActionResult EventLogEntry()
        {
            string source = "MyApp";
            string logName = "Application";
            string userInput = Request.QueryString["data"];

            if (!EventLog.SourceExists(source))
            {
                EventLog.CreateEventSource(source, logName);
            }

            // VULNERABLE: User input in event log
            EventLog.WriteEntry(source, $"User action: {userInput}");
            return Ok();
        }

        // Test 5: log4net logging
        public ActionResult Log4NetLog()
        {
            string userId = Request.QueryString["userId"];
            // VULNERABLE: Log forging via CRLF
            _log4net.Info($"User {userId} accessed resource");
            return Ok();
        }

        // Test 6: NLog logging
        public ActionResult NLogLog()
        {
            string error = Request.Form["error"];
            // VULNERABLE: User-controlled error message
            _nlog.Error($"Error occurred: {error}");
            return Ok();
        }

        // Test 7: Microsoft.Extensions.Logging
        public ActionResult MsLog()
        {
            string data = Request.QueryString["data"];
            // VULNERABLE: User input in structured log
            _msLogger.LogInformation("Request data: {Data}", data);
            return Ok();
        }

        // Test 8: Trace logging
        public ActionResult TraceLog()
        {
            string input = Request.Form["input"];
            // VULNERABLE: User input in trace
            Trace.WriteLine($"Processing: {input}");
            return Ok();
        }

        // Test 9: StreamWriter logging
        public ActionResult StreamLog()
        {
            string audit = Request.Form["audit"];
            using (var writer = new StreamWriter("/var/log/audit.log", true))
            {
                // VULNERABLE: Can break log format
                writer.WriteLine($"[AUDIT] {DateTime.Now}: {audit}");
            }
            return Ok();
        }

        // Test 10: JSON log format breaking
        public ActionResult JsonLog()
        {
            string message = Request.QueryString["msg"];
            // VULNERABLE: Can break JSON structure
            string logEntry = $"{{\"timestamp\":\"{DateTime.Now}\",\"message\":\"{message}\"}}";
            File.AppendAllText("/var/log/json.log", logEntry + "\n");
            return Ok();
        }

        // Test 11: Exception message logging
        public ActionResult ExceptionLog()
        {
            try
            {
                throw new Exception(Request.QueryString["error"]);
            }
            catch (Exception ex)
            {
                // VULNERABLE: Exception message from user input
                Console.WriteLine($"Exception: {ex.Message}");
            }
            return Ok();
        }
    }
}
